Add basic figshare API interaction

vasudevanv opened the pull request:
Added some code with a figshare class and methods to create and update articles. Figuring out how to handle OAuth2.

vasudevanv committed:
Initial commit

Initial commit with code to access the Figshare API. The code is
entirely based on the example in the figshare API documentation.
See  https://docs.figshare.com/api/upload_example/

vasudevanv commented:
Need to address
* Add functionality to download specific versions of articles with version=None referring to the latest version.
* Download article without tokens
* Add some simple function wrappers

dhimmel commented on figshare/figshare.py:
```diff
@@ -0,0 +1,111 @@
+from requests.exceptions import HTTPError
+
+
+def issue_request(method, url, headers, data=None, binary=False):
```
Let's add [docstrings](http://docs.python-guide.org/en/latest/writing/documentation/) describing what the function does and what the parameters are.

dhimmel commented on figshare/figshare.py:
```diff
@@ -0,0 +1,111 @@
+    def update_article(self, article_id, **kwargs):
+        """ Updates an article with a given article_id. """
+        allowed = {'title', 'description', 'defined_type', 'tags', 'categories'}
+        valid_keys = set(kwargs.keys()).intersection(set(allowed))
```
`set(kwargs.keys())` is the same as `set(kwargs)`. `allowed` is already a set.

dhimmel commented on figshare/figshare.py:
```diff
@@ -0,0 +1,111 @@
+        for key in valid_keys:
+            body[key] = kwargs[key]
+
+        url = self.endpoint('/account/articles/{0}'.format(str(article_id)))
```
`str` not needed, format will automatically convert into to string.

dhimmel commented on figshare/figshare.py:
```diff
@@ -0,0 +1,111 @@
+        response = self.issue_request('POST', url, headers=headers, data=data)
+        response = json.loads(response.body_string())
+
+        if "error" not in response.keys():
```
`.keys()` not needed here.

dhimmel commented on figshare/figshare.py:
```diff
@@ -0,0 +1,111 @@
+import sys
+import os
+import hashlib
+import json
```
Newline after last builtin import.

dhimmel reviewed (commented):
Some minor comments. Main thing is documentation of functions.

vasudevanv committed:
Updated Figshare class to access public articles

Updated the class with a "private" attribute. This will allow control
of access to article either in public or private (with a key) mode.
The object needs to be instantiated with a token while creating or
updating an article.
.

vasudevanv committed:
Added README and setup script

Added a README file and python setup script.

vasudevanv commented:
@dhimmel Updated the code based on your comments. Also added a setup script. Going to add more docstrings before merging.

vasudevanv committed:
Added __init__.py

Added blank __init__.py file

vasudevanv committed:
Minor bugfix in request response

The issue_request function already returns the json response text
as a python dict. Need not do json.loads within the Figshare class
methods.

vasudevanv committed:
Improved docstrings

Added more detailed docstrings documenting the Figshare class.

dhimmel reviewed (approved):
@vasudevanv looks great.

I will merge this now. Excited to try it out.
